Mathlib/Analysis/Normed/Field/Basic.lean

```diff
@@ -669,6 +669,46 @@ theorem nndist_inv_inv₀ {z w : α} (hz : z ≠ 0) (hw : w ≠ 0) :
     nndist z⁻¹ w⁻¹ = nndist z w / (‖z‖₊ * ‖w‖₊) :=
   NNReal.eq <| dist_inv_inv₀ hz hw
 
+namespace NormedDivisionRing
+
+section Discrete
+
+variable {𝕜 : Type*} [NormedDivisionRing 𝕜] [DiscreteTopology 𝕜]
```

Comment:
I agree that the proof needs k to be a division ring, but is it really needed mathematically?

Is it also possible to add some comments in the intro of the file about this result?

Reply:
Here's a counterexample of a `NormedRing` that has `DiscreteTopology` but where nonzeros have a norm that isn't 1. The counterexample is the integers, but all the norms have been scaled by 2. This is because `norm_inv` in Mathlib requires `NormedDivisionRing`. Similarly, `norm_one` isn't true for an arbitrary `NormedRing`.

```
import Mathlib.Analysis.Normed.Field.Lemmas

def Z := ℤ
deriving DecidableEq, LinearOrderedCommRing

@[simp]
def Z.toInt (z : Z) : ℤ := z

@[simp] noncomputable def Z.norm (z : Z) : ℝ := 2 • ‖z.toInt‖

noncomputable instance : Norm Z := ⟨Z.norm⟩

noncomputable instance : NormedRing Z where
  dist x y := Z.norm (x - y)
  dist_self := by simp
  dist_comm := by simp [← dist_eq_norm, dist_comm]
  dist_triangle x y z := by
    simp only [Z.norm, Z.toInt, nsmul_eq_mul, Nat.cast_ofNat, ← mul_add]
    simpa using norm_sub_le_norm_sub_add_norm_sub _ _ _
  eq_of_dist_eq_zero := by simp [sub_eq_zero]
  dist_eq _ _ := by congr
  norm_mul x y := by
    simp only [norm, Z.norm, Z.toInt, nsmul_eq_mul, Nat.cast_ofNat, mul_assoc,
                Nat.ofNat_pos, mul_le_mul_left]
    rw [mul_left_comm, ← Real.norm_eq_abs, Int.cast_mul]
    refine (norm_mul_le _ _).trans ?_
    simp [two_mul, mul_nonneg]

@[simp] lemma norm_lt_one {z : Z} : ‖z‖ < 1 ↔ z = 0 := by
  simp only [norm, Z.norm]
  have : 2 • |(z.toInt : ℝ)| = |2 • z.toInt| := by
    simp only [Z.toInt, nsmul_eq_mul, Nat.cast_ofNat]
    rw [← Int.norm_eq_abs]
    simp only [Int.cast_abs, Int.cast_mul, Int.cast_ofNat]
    rw [← Real.norm_eq_abs, norm_mul]
    simp [Int.norm_eq_abs]
  rw [this]
  norm_cast
  simp [Int.abs_lt_one_iff]

instance : DiscreteTopology Z := by
  rw [discreteTopology_iff_nhds]
  intro x
  ext y
  simp only [Metric.mem_nhds_iff, gt_iff_lt, Filter.mem_pure]
  constructor
  · rintro ⟨ε, εpos, h⟩
    exact h (by simp [εpos])
  · intro h
    use 1
    simp only [zero_lt_one, true_and]
    intro z
    simp (config := {contextual := true}) [dist_eq_norm, sub_eq_zero, h]

example {z : Z} : ‖z‖ > 1 ↔ z ≠ 0 := by
  rw [iff_not_comm, ← norm_lt_one, gt_iff_lt, not_lt, lt_iff_le_not_le]
  simp only [not_le, norm_lt_one, and_iff_left_iff_imp]
  simp only [norm, Z.norm, Nat.cast_ofNat]
  have : 2 • |(z.toInt : ℝ)| = |2 • z.toInt| := by
    simp only [Z.toInt, nsmul_eq_mul, Nat.cast_ofNat]
    rw [← Int.norm_eq_abs]
    simp only [Int.cast_abs, Int.cast_mul, Int.cast_ofNat]
    rw [← Real.norm_eq_abs, norm_mul]
    simp [Int.norm_eq_abs]
  rw [this]
  norm_cast
  simp [Int.abs_le_one_iff, Int.mul_eq_one_iff_eq_one_or_neg_one,
    Int.mul_eq_neg_one_iff_eq_one_or_neg_one]
```

Reply:
WOWOW! That's awesome, thanks.